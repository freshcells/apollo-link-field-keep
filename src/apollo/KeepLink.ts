import type { DocumentNode, DirectiveNode } from 'graphql'
import {
  ApolloLink,
  Observable,
  Operation,
  NextLink,
  FetchResult,
} from '@apollo/client'
import { argumentsObjectFromField } from 'apollo-utilities'
import {
  findNodePaths,
  removeDirectivesFromDocument,
  setInObject,
} from './utils'

export const DIRECTIVE = 'keep'

interface DirectiveArguments extends Record<string, any> {
  if?: boolean
  ifFeature?: string
}

export function removeIgnoreSetsFromDocument<T extends DocumentNode>(
  document: T,
  variables: Record<string, any>,
  enabledFeatures: Array<string>
): { modifiedDoc: T; nullFields: Array<Array<string | number>> } {
  const { modifiedDoc, nodesToRemove } = removeDirectivesFromDocument(
    [
      {
        // if the directive should be removed
        test: (directive?: DirectiveNode) => {
          return directive?.name?.value === DIRECTIVE
        },
        // If we should remove the whole field request
        remove: (directive?: DirectiveNode) => {
          if (directive?.name?.value === DIRECTIVE) {
            const args: DirectiveArguments = argumentsObjectFromField(
              directive,
              variables
            )
            if (args.ifFeature !== undefined && args.if !== undefined) {
              return !(
                args.if && enabledFeatures.indexOf(args.ifFeature) !== -1
              )
            }
            if (args.ifFeature) {
              return enabledFeatures.indexOf(args.ifFeature) === -1
            }
            return args.if === false
          }
          return false
        },
      },
    ],
    document
  )

  // The field path's where we have set null later
  const nullFields = findNodePaths(nodesToRemove, document)

  return { modifiedDoc: modifiedDoc as T, nullFields }
}

export class KeepLink extends ApolloLink {
  private _enabledFeatures: string[] = []

  public set enabledFeatures(features: string[]) {
    this._enabledFeatures = features
  }

  constructor(features: string[] = []) {
    super()
    this._enabledFeatures = features
  }

  request(operation: Operation, forward: NextLink): Observable<FetchResult> {
    const directives = `directive @${DIRECTIVE} on FIELD`

    operation.setContext(({ schemas = [] }) => ({
      schemas: (schemas as Record<string, string>[]).concat([{ directives }]),
    }))
    const { modifiedDoc, nullFields } = removeIgnoreSetsFromDocument(
      operation.query,
      operation.variables,
      this._enabledFeatures
    )
    operation.query = modifiedDoc
    // only apply the changes if we actually removed fields.
    if (nullFields.length > 0) {
      return new Observable((observer) => {
        const obs = forward(operation)
        return obs.subscribe({
          next({
            data,
            errors,
          }: FetchResult<Record<string | number, unknown>>) {
            nullFields.forEach((fields) => {
              // To fix issues with apollo caching, we pretend that the server still returns the fields
              // but with null values, this is according to the GraphQL spec.
              setInObject(data, fields, null)
            })
            return observer.next({ data, errors })
          },
          error: observer.error.bind(observer),
          complete: observer.complete.bind(observer),
        })
      })
    }
    return forward(operation)
  }
}
