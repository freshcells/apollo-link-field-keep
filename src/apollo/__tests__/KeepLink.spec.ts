import { gql } from '@apollo/client'
import { print } from 'graphql/language/printer'
import { removeIgnoreSetsFromDocument } from '../KeepLink'
import { setInObject } from '../utils'

describe('A document without `@keep`', () => {
  const queryWithDefaults = gql`
    query someQuery {
      test {
        some
      }
      field {
        subFieldA
        subFieldB
      }
    }
  `
  it('should not modify the document', () => {
    const { modifiedDoc } = removeIgnoreSetsFromDocument(
      queryWithDefaults,
      {},
      []
    )
    expect(modifiedDoc).toBe(queryWithDefaults)
  })
  it('should work with a second time with caches', () => {
    const { modifiedDoc } = removeIgnoreSetsFromDocument(
      queryWithDefaults,
      {},
      []
    )
    expect(modifiedDoc).toBe(queryWithDefaults)
  })
})
describe('KeepLink with if', () => {
  const query = gql`
    query someQuery($shouldKeep: Boolean!) {
      test {
        some
      }
      field @keep(if: $shouldKeep) {
        subFieldA
        subFieldB
      }
    }
  `
  it('should remove directives and fields if shouldKeep is false', () => {
    const { modifiedDoc } = removeIgnoreSetsFromDocument(
      query,
      {
        shouldKeep: false,
      },
      []
    )
    expect(print(modifiedDoc)).toMatchSnapshot()
  })
  it('should keep fields, but remove directives if shouldKeep is true', () => {
    const { modifiedDoc } = removeIgnoreSetsFromDocument(
      query,
      {
        shouldKeep: true,
      },
      []
    )
    expect(print(modifiedDoc)).toMatchSnapshot()
  })

  describe('Field Arguments', () => {
    const queryWithArgument = gql`
      query someQuery($argValue: String!, $shouldKeep: Boolean!) {
        someOther {
          subFieldOther
        }
        field(someArg: $argValue) @keep(if: $shouldKeep) {
          subFieldA
          subFieldB
        }
      }
    `
    const queryWithArgumentAndUsedElsewhere = gql`
      query someQuery($argValue: String!, $shouldKeep: Boolean!) {
        someOther {
          subFieldOther(someArg: $argValue) {
            subSubFieldOther
          }
        }
        field(someArg: $argValue) @keep(if: $shouldKeep) {
          subFieldA
          subFieldB
        }
      }
    `

    it('should remove query argument variables if exists', () => {
      const { modifiedDoc } = removeIgnoreSetsFromDocument(
        queryWithArgument,
        {
          shouldKeep: false,
          argValue: 'test',
        },
        []
      )
      expect(print(modifiedDoc)).toMatchSnapshot()
    })

    it('should keep variables if they are used elsewhere', () => {
      const { modifiedDoc } = removeIgnoreSetsFromDocument(
        queryWithArgumentAndUsedElsewhere,
        { shouldKeep: false, argValue: 'test' },
        []
      )
      expect(print(modifiedDoc)).toMatchSnapshot()
    })
  })

  describe('Deep Fields', () => {
    const deeperQuery = gql`
      query someQuery($argValue: String!, $shouldKeep: Boolean!) {
        someOther {
          subFieldOther
        }
        field {
          subFieldA
          subFieldB {
            deep(someArg: $argValue) @keep(if: $shouldKeep) {
              test {
                wow
              }
            }
          }
        }
      }
    `
    const { modifiedDoc, nullFields } = removeIgnoreSetsFromDocument(
      deeperQuery,
      { shouldKeep: false, argValue: 'test' },
      []
    )
    expect(nullFields).toEqual([['field', 'subFieldB', 'deep']])
    expect(print(modifiedDoc)).toMatchSnapshot()
  })

  describe('Fragment', () => {
    const someFragment = gql`
      fragment ifQueryFragment on Type {
        arg @keep(if: $shouldKeep) {
          test
        }
      }
    `
    const fragmentQuery = gql`
      query someQuery($shouldKeep: Boolean!) {
        deeply {
          nested {
            ...ifQueryFragment
          }
        }
      }
      ${someFragment}
    `

    it('should work with fragments', () => {
      const { modifiedDoc, nullFields } = removeIgnoreSetsFromDocument(
        fragmentQuery,
        { shouldKeep: false },
        []
      )
      expect(nullFields).toEqual([['deeply', 'nested', 'arg']])
      expect(print(modifiedDoc)).toMatchSnapshot()
    })
  })

  describe('Aliases', () => {
    const someFragment = gql`
      fragment someOtherIfFragment on Type {
        aliasForSome: some @keep(if: $shouldKeep) {
          test
        }
      }
    `
    const fragmentQuery = gql`
      query someQuery($shouldKeep: Boolean!) {
        aliasForDeeply: deeply {
          nested {
            ...someOtherIfFragment
          }
        }
      }
      ${someFragment}
    `

    it('should properly map aliases', () => {
      const { modifiedDoc, nullFields } = removeIgnoreSetsFromDocument(
        fragmentQuery,
        { shouldKeep: false },
        []
      )
      expect(nullFields).toEqual([['aliasForDeeply', 'nested', 'aliasForSome']])
      expect(print(modifiedDoc)).toMatchSnapshot()
    })
  })

  describe('Empty documents', () => {
    const queryThatWillBeEmpty = gql`
      query someQuery($argValue: String!, $shouldKeep: Boolean!) {
        field(someArg: $argValue) @keep(if: $shouldKeep) {
          subFieldA
          subFieldB
        }
      }
    `
    const { modifiedDoc, nullFields } = removeIgnoreSetsFromDocument(
      queryThatWillBeEmpty,
      { shouldKeep: false },
      []
    )

    expect(nullFields).toEqual([['field']])
    expect(modifiedDoc).toBe(null)
  })
})

describe('KeepLink with ifFeature', () => {
  const features = ['feature1']
  const query = gql`
    query someQuery {
      test {
        some
      }
      field @keep(ifFeature: "feature1") {
        subFieldA
        subFieldB
      }
    }
  `
  it('should remove directives and fields if feature1 is not enabled', () => {
    const { modifiedDoc } = removeIgnoreSetsFromDocument(query, {}, [])
    expect(print(modifiedDoc)).toMatchSnapshot()
  })
  it('should keep fields, but remove directives if feature1 is enabled', () => {
    const { modifiedDoc } = removeIgnoreSetsFromDocument(query, {}, features)
    expect(print(modifiedDoc)).toMatchSnapshot()
  })

  describe('Field Arguments', () => {
    const queryWithArgument = gql`
      query someQuery($argValue: String!) {
        someOther {
          subFieldOther
        }
        field(someArg: $argValue) @keep(ifFeature: "feature1") {
          subFieldA
          subFieldB
        }
      }
    `
    const queryWithArgumentAndUsedElsewhere = gql`
      query someQuery($argValue: String!) {
        someOther {
          subFieldOther(someArg: $argValue) {
            subSubFieldOther
          }
        }
        field(someArg: $argValue) @keep(ifFeature: "feature1") {
          subFieldA
          subFieldB
        }
      }
    `

    it('should remove argument variables if exists', () => {
      const { modifiedDoc } = removeIgnoreSetsFromDocument(
        queryWithArgument,
        {
          argValue: 'test',
        },
        []
      )
      expect(print(modifiedDoc)).toMatchSnapshot()
    })
    it('should keep variables if they are used elsewhere', () => {
      const { modifiedDoc } = removeIgnoreSetsFromDocument(
        queryWithArgumentAndUsedElsewhere,
        { argValue: 'test' },
        []
      )
      expect(print(modifiedDoc)).toMatchSnapshot()
    })
  })

  describe('Deep Fields', () => {
    const deeperQuery = gql`
      query someQuery($argValue: String!) {
        someOther {
          subFieldOther
        }
        field {
          subFieldA
          subFieldB {
            deep(someArg: $argValue) @keep(ifFeature: "feature1") {
              test {
                wow
              }
            }
          }
        }
      }
    `
    const { modifiedDoc, nullFields } = removeIgnoreSetsFromDocument(
      deeperQuery,
      { argValue: 'test' },
      []
    )
    expect(nullFields).toEqual([['field', 'subFieldB', 'deep']])
    expect(print(modifiedDoc)).toMatchSnapshot()
  })

  describe('Fragment', () => {
    const someFragment = gql`
      fragment ifFeatureQueryFragment on Type {
        arg @keep(ifFeature: "feature1") {
          test
        }
      }
    `
    const fragmentQuery = gql`
      query someQuery {
        deeply {
          nested {
            ...ifFeatureQueryFragment
          }
        }
      }
      ${someFragment}
    `

    it('should work with fragments', () => {
      const { modifiedDoc, nullFields } = removeIgnoreSetsFromDocument(
        fragmentQuery,
        {},
        []
      )
      expect(nullFields).toEqual([['deeply', 'nested', 'arg']])
      expect(print(modifiedDoc)).toMatchSnapshot()
    })
  })

  describe('Aliases', () => {
    const someFragment = gql`
      fragment someOtherIfFeatureFragment on Type {
        aliasForSome: some @keep(ifFeature: "feature1") {
          test
        }
      }
    `
    const fragmentQuery = gql`
      query someQuery {
        aliasForDeeply: deeply {
          nested {
            ...someOtherIfFeatureFragment
          }
        }
      }
      ${someFragment}
    `

    it('should properly map aliases', () => {
      const { modifiedDoc, nullFields } = removeIgnoreSetsFromDocument(
        fragmentQuery,
        {},
        []
      )
      expect(nullFields).toEqual([['aliasForDeeply', 'nested', 'aliasForSome']])
      expect(print(modifiedDoc)).toMatchSnapshot()
    })
  })

  describe('Empty documents', () => {
    const queryThatWillBeEmpty = gql`
      query someQuery($argValue: String!, $shouldKeep: Boolean!) {
        field(someArg: $argValue) @keep(ifFeature: "feature1") {
          subFieldA
          subFieldB
        }
      }
    `
    const { modifiedDoc, nullFields } = removeIgnoreSetsFromDocument(
      queryThatWillBeEmpty,
      {},
      []
    )

    expect(nullFields).toEqual([['field']])
    expect(modifiedDoc).toBe(null)
  })
})

describe('KeepLink with if and ifFeature', () => {
  const features = ['feature1']
  const query = gql`
    query someQuery($shouldKeep: Boolean!) {
      test {
        some
      }
      field @keep(if: $shouldKeep, ifFeature: "feature1") {
        subFieldA
        subFieldB
      }
    }
  `
  it('should remove directives and fields if shouldKeep is false and feature1 is not enabled', () => {
    const { modifiedDoc } = removeIgnoreSetsFromDocument(
      query,
      {
        shouldKeep: false,
      },
      []
    )
    expect(print(modifiedDoc)).toMatchSnapshot()
  })
  it('should keep fields, but remove directives if shouldKeep is true and feature1 is enabled', () => {
    const { modifiedDoc } = removeIgnoreSetsFromDocument(
      query,
      {
        shouldKeep: true,
      },
      features
    )
    expect(print(modifiedDoc)).toMatchSnapshot()
  })

  describe('Field Arguments', () => {
    const queryWithArgument = gql`
      query someQuery($argValue: String!, $shouldKeep: Boolean!) {
        someOther {
          subFieldOther
        }
        field(someArg: $argValue)
          @keep(if: $shouldKeep, ifFeature: "feature1") {
          subFieldA
          subFieldB
        }
      }
    `
    const queryWithArgumentAndUsedElsewhere = gql`
      query someQuery($argValue: String!, $shouldKeep: Boolean!) {
        someOther {
          subFieldOther(someArg: $argValue) {
            subSubFieldOther
          }
        }
        field(someArg: $argValue)
          @keep(if: $shouldKeep, ifFeature: "feature1") {
          subFieldA
          subFieldB
        }
      }
    `

    it('should remove argument variables if exists', () => {
      const { modifiedDoc } = removeIgnoreSetsFromDocument(
        queryWithArgument,
        {
          shouldKeep: false,
          argValue: 'test',
        },
        []
      )
      expect(print(modifiedDoc)).toMatchSnapshot()
    })
    it('should keep variables if they are used elsewhere', () => {
      const { modifiedDoc } = removeIgnoreSetsFromDocument(
        queryWithArgumentAndUsedElsewhere,
        { shouldKeep: false, argValue: 'test' },
        []
      )
      expect(print(modifiedDoc)).toMatchSnapshot()
    })
  })

  describe('Deep Fields', () => {
    const deeperQuery = gql`
      query someQuery($argValue: String!, $shouldKeep: Boolean!) {
        someOther {
          subFieldOther
        }
        field {
          subFieldA
          subFieldB {
            deep(someArg: $argValue)
              @keep(if: $shouldKeep, ifFeature: "feature1") {
              test {
                wow
              }
            }
          }
        }
      }
    `
    const { modifiedDoc, nullFields } = removeIgnoreSetsFromDocument(
      deeperQuery,
      { shouldKeep: false, argValue: 'test' },
      []
    )
    expect(nullFields).toEqual([['field', 'subFieldB', 'deep']])
    expect(print(modifiedDoc)).toMatchSnapshot()
  })

  describe('Fragment', () => {
    const someFragment = gql`
      fragment ifAndIfFeatureQueryFragment on Type {
        arg @keep(if: $shouldKeep, ifFeature: "feature1") {
          test
        }
      }
    `
    const fragmentQuery = gql`
      query someQuery($shouldKeep: Boolean!) {
        deeply {
          nested {
            ...ifAndIfFeatureQueryFragment
          }
        }
      }
      ${someFragment}
    `

    it('should work with fragments', () => {
      const { modifiedDoc, nullFields } = removeIgnoreSetsFromDocument(
        fragmentQuery,
        { shouldKeep: false },
        []
      )
      expect(nullFields).toEqual([['deeply', 'nested', 'arg']])
      expect(print(modifiedDoc)).toMatchSnapshot()
    })
  })

  describe('Aliases', () => {
    const someFragment = gql`
      fragment someOtherIfAndIfFeatureFragment on Type {
        aliasForSome: some @keep(if: $shouldKeep, ifFeature: "feature1") {
          test
        }
      }
    `
    const fragmentQuery = gql`
      query someQuery($shouldKeep: Boolean!) {
        aliasForDeeply: deeply {
          nested {
            ...someOtherIfAndIfFeatureFragment
          }
        }
      }
      ${someFragment}
    `

    it('should properly map aliases', () => {
      const { modifiedDoc, nullFields } = removeIgnoreSetsFromDocument(
        fragmentQuery,
        { shouldKeep: false },
        []
      )
      expect(nullFields).toEqual([['aliasForDeeply', 'nested', 'aliasForSome']])
      expect(print(modifiedDoc)).toMatchSnapshot()
    })
  })

  describe('Empty documents', () => {
    const queryThatWillBeEmpty = gql`
      query someQuery($argValue: String!, $shouldKeep: Boolean!) {
        field(someArg: $argValue)
          @keep(if: $shouldKeep, ifFeature: "feature1") {
          subFieldA
          subFieldB
        }
      }
    `
    const { modifiedDoc, nullFields } = removeIgnoreSetsFromDocument(
      queryThatWillBeEmpty,
      { shouldKeep: false },
      []
    )

    expect(nullFields).toEqual([['field']])
    expect(modifiedDoc).toBe(null)
  })
})

describe('utilities', () => {
  describe('setInObject', () => {
    it('should work for full objects', () => {
      const object: Record<string, any> = {
        a: [
          { b: { wow: 'test' } },
          { b: { wow: 'test' } },
          { b: { wow: 'test' } },
        ],
      }
      setInObject(object, ['a', 'b', 'nice'], null)
      expect(object).toEqual({
        a: [
          { b: { wow: 'test', nice: null } },
          { b: { wow: 'test', nice: null } },
          {
            b: {
              wow: 'test',
              nice: null,
            },
          },
        ],
      })
    })
    it('should work for null paths', () => {
      const object: Record<string, any> = { a: null }
      setInObject(object, ['a', 'b', 'nice'], null)
      expect(object).toEqual({ a: null })
    })
    it('should work for undefined paths', () => {
      const object: Record<string, any> = {}
      setInObject(object, ['a', 'b', 'nice'], null)
      expect(object).toEqual({})
    })
  })
})

describe('fragments', () => {
  it('should remove unused fragments', () => {
    const queryWithFragments = gql`
      fragment myFragment on SomeType {
        field @keep(ifFeature: "someFeature")
      }
      query myQuery {
        field {
          someOtherKey
          ...myFragment
        }
      }
    `
    const { nullFields, modifiedDoc } = removeIgnoreSetsFromDocument(
      queryWithFragments,
      {},
      []
    )
    expect(nullFields).toEqual([['field', 'field']])
    expect(print(modifiedDoc)).toEqual(
      print(gql`
        query myQuery {
          field {
            someOtherKey
          }
        }
      `)
    )
  })
})

describe('complex nested mutations and fragments', () => {
  const complexNested = gql`
    fragment level2 on OtherType {
      nice
      to
      have @keep(ifFeature: "someFeature")
      some @keep(ifFeature: "someFeature") {
        subselection
      }
    }
    fragment level1 on MyType {
      label
      oneOnLevel1 @keep(ifFeature: "someFeature")
      goesToLevel2 {
        ...level2
      }
    }
    mutation complexWithNestedFragments {
      complexWithNestedFragments {
        nested {
          firstLevelKey @keep(ifFeature: "someFeature")
          ...level1
        }
      }
    }
  `

  const { nullFields, modifiedDoc } = removeIgnoreSetsFromDocument(
    complexNested,
    {},
    []
  )
  expect(nullFields).toEqual([
    ['complexWithNestedFragments', 'nested', 'goesToLevel2', 'some'],
    ['complexWithNestedFragments', 'nested', 'goesToLevel2', 'have'],
    ['complexWithNestedFragments', 'nested', 'oneOnLevel1'],
    ['complexWithNestedFragments', 'nested', 'firstLevelKey'],
  ])
  expect(print(modifiedDoc)).toMatchSnapshot()
})
describe('handle inline fragments', () => {
  const inlineFragment = gql`
    mutation complexWithNestedFragments {
      complexWithNestedFragments {
        nested {
          ... on SomeType {
            firstLevelKey @keep(ifFeature: "someFeature")
            primaryFunction
            ... on SomeOtherType {
              deeper @keep(ifFeature: "someFeature")
            }
          }
        }
      }
    }
  `
  const { nullFields, modifiedDoc } = removeIgnoreSetsFromDocument(
    inlineFragment,
    {},
    []
  )
  expect(nullFields).toEqual([
    ['complexWithNestedFragments', 'nested', 'deeper'],
    ['complexWithNestedFragments', 'nested', 'firstLevelKey'],
  ])
  expect(print(modifiedDoc)).toMatchSnapshot()
})
