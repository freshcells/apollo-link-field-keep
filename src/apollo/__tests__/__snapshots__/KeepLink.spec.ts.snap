// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
"query someQuery {
  someOther {
    subFieldOther
  }
  field {
    subFieldA
    subFieldB
  }
}
"
`;

exports[` 2`] = `
"query someQuery {
  someOther {
    subFieldOther
  }
  field {
    subFieldA
    subFieldB
  }
}
"
`;

exports[` 3`] = `
"query someQuery {
  someOther {
    subFieldOther
  }
  field {
    subFieldA
    subFieldB
  }
}
"
`;

exports[` 4`] = `
"fragment level2 on OtherType {
  nice
  to
}

fragment level1 on MyType {
  label
  goesToLevel2 {
    ...level2
  }
}

mutation complexWithNestedFragments {
  complexWithNestedFragments {
    nested {
      ...level1
    }
  }
}
"
`;

exports[` 5`] = `
"mutation complexWithNestedFragments {
  complexWithNestedFragments {
    nested {
      ... on SomeType {
        primaryFunction
      }
    }
  }
}
"
`;

exports[`Default Variables should handle multiple default variables correctly 1`] = `
"query someQuery($otherVar: String = \\"default\\") {
  test {
    some
  }
  otherField(arg: $otherVar) {
    data
  }
}
"
`;

exports[`Default Variables should override default value when variable is explicitly provided 1`] = `
"query someQuery {
  test {
    some
  }
}
"
`;

exports[`Default Variables should use default value when variable is not provided 1`] = `
"query someQuery {
  test {
    some
  }
  field {
    subFieldA
    subFieldB
  }
}
"
`;

exports[`KeepLink with if Aliases should properly map aliases 1`] = `
"query someQuery {
  aliasForDeeply: deeply {
    nested
  }
}
"
`;

exports[`KeepLink with if Field Arguments should keep variables if they are used elsewhere 1`] = `
"query someQuery($argValue: String!) {
  someOther {
    subFieldOther(someArg: $argValue) {
      subSubFieldOther
    }
  }
}
"
`;

exports[`KeepLink with if Field Arguments should remove query argument variables if exists 1`] = `
"query someQuery {
  someOther {
    subFieldOther
  }
}
"
`;

exports[`KeepLink with if Fragment should work with fragments 1`] = `
"query someQuery {
  deeply {
    nested
  }
}
"
`;

exports[`KeepLink with if and ifFeature Aliases should properly map aliases 1`] = `
"query someQuery {
  aliasForDeeply: deeply {
    nested
  }
}
"
`;

exports[`KeepLink with if and ifFeature Field Arguments should keep variables if they are used elsewhere 1`] = `
"query someQuery($argValue: String!) {
  someOther {
    subFieldOther(someArg: $argValue) {
      subSubFieldOther
    }
  }
}
"
`;

exports[`KeepLink with if and ifFeature Field Arguments should remove argument variables if exists 1`] = `
"query someQuery {
  someOther {
    subFieldOther
  }
}
"
`;

exports[`KeepLink with if and ifFeature Fragment should work with fragments 1`] = `
"query someQuery {
  deeply {
    nested
  }
}
"
`;

exports[`KeepLink with if and ifFeature should keep fields, but remove directives if shouldKeep is true and feature1 is enabled 1`] = `
"query someQuery {
  test {
    some
  }
  field {
    subFieldA
    subFieldB
  }
}
"
`;

exports[`KeepLink with if and ifFeature should remove directives and fields if shouldKeep is false and feature1 is not enabled 1`] = `
"query someQuery {
  test {
    some
  }
}
"
`;

exports[`KeepLink with if should keep fields, but remove directives if shouldKeep is true 1`] = `
"query someQuery {
  test {
    some
  }
  field {
    subFieldA
    subFieldB
  }
}
"
`;

exports[`KeepLink with if should remove directives and fields if shouldKeep is false 1`] = `
"query someQuery {
  test {
    some
  }
}
"
`;

exports[`KeepLink with ifFeature Aliases should properly map aliases 1`] = `
"query someQuery {
  aliasForDeeply: deeply {
    nested
  }
}
"
`;

exports[`KeepLink with ifFeature Field Arguments should keep variables if they are used elsewhere 1`] = `
"query someQuery($argValue: String!) {
  someOther {
    subFieldOther(someArg: $argValue) {
      subSubFieldOther
    }
  }
}
"
`;

exports[`KeepLink with ifFeature Field Arguments should remove argument variables if exists 1`] = `
"query someQuery {
  someOther {
    subFieldOther
  }
}
"
`;

exports[`KeepLink with ifFeature Fragment should work with fragments 1`] = `
"query someQuery {
  deeply {
    nested
  }
}
"
`;

exports[`KeepLink with ifFeature should keep fields, but remove directives if feature1 is enabled 1`] = `
"query someQuery {
  test {
    some
  }
  field {
    subFieldA
    subFieldB
  }
}
"
`;

exports[`KeepLink with ifFeature should remove directives and fields if feature1 is not enabled 1`] = `
"query someQuery {
  test {
    some
  }
}
"
`;
